x/milestone/abci/abci.go

```diff
@@ -0,0 +1,173 @@
+package abci
+
+import (
+	"fmt"
+	"math/big"
+	"sort"
+	"time"
+
+	"cosmossdk.io/log"
+
+	"github.com/0xPolygon/heimdall-v2/helper"
+	"github.com/0xPolygon/heimdall-v2/sidetxs"
+	"github.com/0xPolygon/heimdall-v2/x/milestone/keeper"
+	"github.com/0xPolygon/heimdall-v2/x/milestone/types"
+	stakeTypes "github.com/0xPolygon/heimdall-v2/x/stake/types"
+	abciTypes "github.com/cometbft/cometbft/abci/types"
+	cmtTypes "github.com/cometbft/cometbft/proto/tendermint/types"
+	"github.com/cosmos/cosmos-sdk/codec/address"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+var pendingMilestoneProposition *sidetxs.MilestoneProposition
+
+func GenMilestoneProposition(ctx sdk.Context, milestoneKeeper *keeper.Keeper, contractCaller helper.IContractCaller) (*sidetxs.MilestoneProposition, error) {
+	propStartBlock := uint64(0)
+
+	milestone, err := milestoneKeeper.GetLastMilestone(ctx)
+	if err != nil && err != types.ErrNoMilestoneFound {
+		return nil, err
+	}
+
+	pendingMilestone := GetPendingMilestoneProposition()
+
+	logger := ctx.Logger()
+
+	timeSinceLastMilestone := time.Duration(0)
+
+	if milestone != nil {
+		timeSinceLastMilestone = time.Since(time.Unix(int64(milestone.Timestamp), 0))
```

Comment:
I am not sure about using time.Since, I think we can have a case where the check passes on some nodes and not on other. I think better to depend on req.Time, which should be the time set by the block proposer.

Reply:
Great point, will check if I can use req time here.

Reply:
But instead of checking for time, isn't it more reliable to save at which height the milestone was proposed/finalized and at what height we are now, so we can for sure know that X number of blocks have passed ? Because time check is a bit iffy, there is no guarantee that 6 seconds are 3 blocks because block time is a bit elastic, we can have multiple rounds per block sometimes.

Reply:
Yeah, checking block height might be even better.